### PR TITLE
Support thematic breaks

### PR DESCRIPTION
**Pull request by emilyychenn**
Bug/issue #: rdar://124275218

## Summary

Markdown supports thematic breaks (aka horizontal rules) that are represented by placing three or more hyphens, asterisks, or underscores on a separate line, surrounded by blank lines. This PR adds support for thematic breaks within Swift DocC, adding a `thematicBreak` element in the render JSON.

[Minor fix]: also fixed a consistent typo in RenderContentCompilerTests.

## Dependencies

DocC render needs to add support in order to render this thematic break from the render JSON.

## Testing

Create a thematic break by add three or more hyphens, asterisks, or underscores on a separate line, surrounded by blank lines, and compile the documentation. The render JSON for the data type should now include an element of type "thematicBreak".

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [x] Added tests
- [x] Ran the `./bin/test` script and it succeeded
- [x] Updated documentation if necessary - N/A

### CODE REVIEW

File: Tests/SwiftDocCTests/Rendering/RenderBlockContent_ThematicBreakTests.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        let thematicBreak = RenderBlockContent.thematicBreak
+        let data = try JSONEncoder().encode(thematicBreak)
+        let roundtrip = try JSONDecoder().decode(RenderBlockContent.self, from: data)
+        
+        guard case RenderBlockContent.thematicBreak = roundtrip else {
+            XCTFail("Didn't decode thematic break correctly.")
+            return
+        }
+        
+        XCTAssertEqual(thematicBreak, roundtrip)
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Rendering/RenderBlockContent_ThematicBreakTests.swift:28]: FYI: We have a `assertRoundTripCoding()` test helper for this. 
```suggestion
        assertRoundTripCoding(RenderBlockContent.thematicBreak)
```

File: Sources/SwiftDocC/Model/Rendering/Content/RenderBlockContent.swift [hunk 2]
```diff
@@ -76,7 +76,10 @@ public enum RenderBlockContent: Equatable {
     
     /// A video with an optional caption.
     case video(Video)
-
+    
+    /// A horizontal rule.
```
Comment (d-ronnqvist) [inline Sources/SwiftDocC/Model/Rendering/Content/RenderBlockContent.swift:80]: If I understand correctly the "horizontal rule" terminology is somewhat specific to the HTML element and "thematic break" is a more abstract term that describe the use of this element. 

```suggestion
    /// An authored thematic break between block elements.
```

File: Tests/SwiftDocCTests/Model/SemaToRenderNodeTests.swift [hunk 3]
```diff
@@ -3452,4 +3452,23 @@ Document
...
+    
+    func testThematicBreak() throws {
+        let source = """
+
+        ---
+
+        """
+        
+        let markup = Document(parsing: source, options: .parseBlockDirectives)
+        let (bundle, context) = try testBundleAndContext(named: "TestBundle")
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Model/SemaToRenderNodeTests.swift:3464]: Would `testBundleAndContext()` work here? If possible it would be good not build the test bundle only to discard it.

File: Tests/SwiftDocCTests/Rendering/RenderContentCompilerTests.swift [hunk 4]
```diff
@@ -181,19 +181,46 @@ class RenderContentCompilerTests: XCTestCase {
...
+                XCTFail("RenderContent result is not the expected RenderBlockContent.paragraph(Paragraph)")
                 return
             }
             let text = RenderInlineContent.text("\n")
             XCTAssertEqual(paragraph.inlineContent[1], text)
         }
     }
+    
+    func testThematicBreak() throws {
+        let (bundle, context) = try testBundleAndContext(named: "TestBundle")
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Rendering/RenderContentCompilerTests.swift:201]: Like in the other test, is is possible to use an empty bundle here instead of building and discarding the content?

File: Tests/SwiftDocCTests/Model/SemaToRenderNodeTests.swift [hunk 5]
```diff
@@ -3452,4 +3452,23 @@ Document
...
+        var contentTranslator = RenderContentCompiler(context: context, bundle: bundle, identifier: ResolvedTopicReference(bundleIdentifier: bundle.identifier, path: "/TestTutorial", sourceLanguage: .swift))
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Model/SemaToRenderNodeTests.swift:3465]: minor: This path doesn't seem related to the content.

File: Sources/SwiftDocC/Model/Rendering/Content/RenderBlockContent.swift [hunk 6]
```diff
@@ -76,7 +76,10 @@ public enum RenderBlockContent: Equatable {
...
+    case thematicBreak
```
Comment (d-ronnqvist) [inline Sources/SwiftDocC/Model/Rendering/Content/RenderBlockContent.swift:81]: I think we need a new entry in the render node spec for this (which would also bump the patch component of the version):

```json
"ThematicBreak": {
    "required": [
        "type",
    ],
    "type": "object",
    "properties": {
        "type": {
            "type": "string",
            "enum": [
                "thematicBreak"
            ]
        }
    }
},
```

File: Tests/SwiftDocCTests/Model/SemaToRenderNodeTests.swift [hunk 7]
```diff
@@ -3452,4 +3452,23 @@ Document
             ])
         }
     }
+    
+    func testThematicBreak() throws {
+        let source = """
+
+        ---
```
Comment (patshaughnessy) [inline Tests/SwiftDocCTests/Model/SemaToRenderNodeTests.swift:3459]: I know Swift-Markdown handles the parsing of thematic breaks, but would it be worth it to add a couple of tests that use `===` and `___` the other two styles you mentioned in this PR?

**Review by patshaughnessy — APPROVED** [review]
Works perfectly for me 👍  - although as you say DocC Render doesn't support this yet but I see the same block content type appear in Render JSON.

File: Tests/SwiftDocCTests/Rendering/RenderContentCompilerTests.swift [hunk 8]
```diff
@@ -62,7 +62,7 @@ class RenderContentCompilerTests: XCTestCase {
         
         do {
             guard case let .paragraph(paragraph) = result[0] as? RenderBlockContent else {
-                XCTFail("RenderCotent result is not the expected RenderBlockContent.paragraph(Paragraph)")
+                XCTFail("RenderContent result is not the expected RenderBlockContent.paragraph(Paragraph)")
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Rendering/RenderContentCompilerTests.swift:65]: Thanks for fixing these typos 👍

File: Tests/SwiftDocCTests/Model/SemaToRenderNodeTests.swift (hunk 5 above)
Comment (emilyychenn) [reply Tests/SwiftDocCTests/Model/SemaToRenderNodeTests.swift:3465]: Renamed! 😅

File: Tests/SwiftDocCTests/Rendering/RenderBlockContent_ThematicBreakTests.swift (hunk 1 above)
Comment (emilyychenn) [reply Tests/SwiftDocCTests/Rendering/RenderBlockContent_ThematicBreakTests.swift:28]: Wow this feels so much cleaner, thanks!

**Review by sofiaromorales — COMMENTED** [review]
Thanks for this 😄 A minor note - when we make changes to a file, we also need to update the year in the license header, like this: `202X-2024`, besides that, this pr looks good to me.

**Comment by mportiz08** [comment]
> Works perfectly for me 👍 - although as you say DocC Render doesn't support this yet but I see the same block content type appear in Render JSON.

It does now as of this PR: https://github.com/apple/swift-docc-render/pull/791 :)

**Review by d-ronnqvist — APPROVED** [review]
LGTM

**Comment by emilyychenn** [comment]
@swift-ci Please test

**Comment by emilyychenn** [comment]
@swift-ci please test
